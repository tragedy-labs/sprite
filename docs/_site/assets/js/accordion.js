/*
 *   This content is licensed according to the W3C Software License at
 *   https://www.w3.org/copyright/software-license-2023/
 *
 *   Simple accordion pattern example
 */

'use strict';

class Accordion {
  constructor(domNode) {
    this.root = domNode;
    this.button = this.root.querySelector('button[aria-expanded]');
    this.icon = this.root.querySelector('svg');
    const controlsId = this.button.getAttribute('aria-controls');
    this.content = document.getElementById(controlsId);

    this.open = this.button.getAttribute('aria-expanded') === 'true';

    // add event listeners
    this.button.addEventListener('click', this.onButtonClick.bind(this));
    this.toggle(false);
  }

  onButtonClick() {
    this.toggle(!this.open);
  }

  toggle(open) {
    // don't do anything if the open state doesn't change
    if (open === this.open) {
      return;
    }

    // update the internal state
    this.open = open;

    // handle DOM updates
    this.button.setAttribute('aria-expanded', `${open}`);
    if (open) {
      this.content.removeAttribute('hidden');
      this.icon.classList.add('open');
    } else {
      this.content.setAttribute('hidden', '');
      this.icon.classList.remove('open');
    }
  }

  // Add public open and close methods for convenience
  open() {
    this.toggle(true);
  }

  close() {
    this.toggle(false);
  }
}

const initializeAccordions = () => {
  // init accordions
  const accordions = document.querySelectorAll('.menuItem');
  accordions.forEach((accordionEl) => {
    new Accordion(accordionEl);
  });
};

document.addEventListener('DOMContentLoaded', initializeAccordions);
