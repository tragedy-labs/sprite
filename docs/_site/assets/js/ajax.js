// Constants
const CONTENT_CONTAINER_ID = "content";
const PARSER = new DOMParser();

// State
let historyStack = [];
let currentStateIndex = -1;

// Functions
async function fetchDocument(url) {
  try {
    const response = await fetch(url);
    return await response.text();
  } catch (error) {
    throw new Error(`Failed to fetch document: ${url}`, { cause: error });
  }
}

async function parseHtml(html) {
  return PARSER.parseFromString(html, "text/html");
}

async function updateContent(htmlDoc) {
  const title = htmlDoc.querySelector("title").textContent;
  const contentDiv = htmlDoc.getElementById(CONTENT_CONTAINER_ID);
  const target = document.getElementById(CONTENT_CONTAINER_ID);
  target.innerHTML = contentDiv.innerHTML;
  target.scrollTop = 0;
  document.title = title;

  // Reinitialize link click events for newly loaded content
  createLinkClickEvents(target);
}

function pushToHistory(url) {
  historyStack = historyStack.slice(0, currentStateIndex + 1);
  historyStack.push(url);
  currentStateIndex++;
  history.pushState({ index: currentStateIndex }, "", url);
}

async function navigate(url, addToHistory = true) {
  const urlObj = new URL(url, window.location.origin);

  if (urlObj)
    if (urlObj.hash && urlObj.pathname === window.location.pathname) {
      // Handle local page anchor navigation
      const anchor = document.querySelector(urlObj.hash);
      if (anchor) {
        anchor.scrollIntoView();
        addToHistory && pushToHistory(url);
        return;
      }
    }

  // Handle Local Navigation
  const html = await fetchDocument(url);
  const htmlDoc = await parseHtml(html);
  await updateContent(htmlDoc);

  addToHistory && pushToHistory(url);
}

function createLinkClickEvents(root) {
  const links = root.querySelectorAll("a");
  links.forEach((link) => {
    link.addEventListener("click", (e) => {
      const url = e.currentTarget.href;
      if (isLocalLink(url)) {
        e.preventDefault();
        navigate(url);
      }
    });
  });
}

function isLocalLink(url) {
  const linkUrl = new URL(url, window.location.origin);
  return linkUrl.origin === window.location.origin;
}

window.addEventListener("popstate", (e) => {
  if (e.state && e.state.index !== undefined) {
    currentStateIndex = e.state.index;
    navigate(historyStack[currentStateIndex], false);
  }
});

// Initialize
function initAjax() {
  createLinkClickEvents(document);
  // Initialize the first state
  historyStack.push(window.location.href);
  currentStateIndex++;
  history.replaceState({ index: currentStateIndex }, "", window.location.href);
}

// Call init
document.addEventListener("DOMContentLoaded", initAjax);
