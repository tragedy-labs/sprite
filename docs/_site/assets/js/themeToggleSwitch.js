function initializeThemeToggle() {
  const element = document.getElementById("theme_toggle");
  const theme = localStorage.getItem("theme");

  const prefersTheme = (theme) =>
    window.matchMedia(`(prefers-color-scheme: ${theme})`).matches;

  const setTheme = (theme) => {
    localStorage.setItem("theme", theme);
    document.body.className = theme;
    setSwitchState(theme);
  };

  const setSwitchState = (theme) => {
    switch (theme) {
      case "dark":
        element.checked = true;
        break;
      case "light":
        element.checked = false;
        break;
      default:
        throw new Error("unknown theme");
    }
  };

  const determinePosition = () => {
    if (theme) {
      setTheme(theme);
    } else {
      if (prefersTheme("dark")) {
        setTheme("dark");
        return;
      } else {
        setTheme("light");
      }
    }
  };

  element.addEventListener("click", () => {
    if (element.checked) {
      setTheme("dark");
    } else {
      setTheme("light");
    }
  });

  determinePosition();
}

document.addEventListener("DOMContentLoaded", initializeThemeToggle);
