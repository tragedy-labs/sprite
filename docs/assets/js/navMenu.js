document.addEventListener("DOMContentLoaded", function () {
  const content = document.getElementById("content");
  const mainNav = document.getElementById("main_nav");
  const navButton = document.getElementById("nav_button");
  const navIndicator = navButton.querySelector("svg");
  const links = mainNav.querySelectorAll("a");

  const closeIconMarkup = `
    <line x1="3" y1="4" x2="21" y2="21"></line>
    <line x1="21" y1="4" x2="3" y2="21"></line>
  `;

  const openIconMarkup = `
    <line x1="3" y1="4" x2="21" y2="4"></line>
    <line x1="3" y1="12" x2="21" y2="12"></line>
    <line x1="3" y1="20" x2="21" y2="20"></line>
  `;

  let isOpen = false;
  let currentTranslation = 0;
  let previousTouchPosition = null;

  function handleTouchEnd() {
    previousTouchPosition = null;

    if (currentTranslation < -80) {
      closeMenu();
    }
    currentTranslation = 0;
    mainNav.style.transform = null;
  }

  function handleTouchStart(event) {
    previousTouchPosition = event.touches[0].clientX;
  }

  function handleTouchMove(event) {
    const currentTouchPosition = event.touches[0].clientX;
    const delta = currentTouchPosition - previousTouchPosition;
    previousTouchPosition = currentTouchPosition;

    currentTranslation += delta;

    if (currentTranslation > 0) {
      currentTranslation = 0;
    }

    mainNav.style.transform = `translate3d(${currentTranslation}px, 0, 0)`;

    const mainNavWidth = mainNav.offsetWidth;
    const ratio = (mainNavWidth + currentTranslation) / mainNavWidth;

    content.style.opacity = 0.65 * (1 - ratio) + 0.35;
  }

  function escHandler(e) {
    if (e.key === "Escape") {
      closeMenu();
    }
  }

  function enableTouchEvents() {
    mainNav.addEventListener("touchstart", handleTouchStart);
    mainNav.addEventListener("touchmove", handleTouchMove);
    mainNav.addEventListener("touchend", handleTouchEnd);
    content.addEventListener("touchstart", closeMenu);
    content.addEventListener("mousedown", closeMenu);
    document.addEventListener("keydown", escHandler);
    // TODO: This could be handled in the ajax script
    // since it already handles link clicks
    links.forEach((link) => {
      link.addEventListener("click", closeMenu);
    });
  }

  function disableTouchEvents() {
    mainNav.removeEventListener("touchstart", handleTouchStart);
    mainNav.removeEventListener("touchmove", handleTouchMove);
    mainNav.removeEventListener("touchend", handleTouchEnd);
    content.removeEventListener("touchstart", closeMenu);
    content.removeEventListener("mousedown", closeMenu);
    document.removeEventListener("keydown", escHandler);
    links.forEach((link) => {
      link.removeEventListener("click", closeMenu);
    });
  }

  function closeMenu() {
    toggleNavState();
    disableTouchEvents();
    navIndicator.innerHTML = openIconMarkup;
    content.style.opacity = null;
  }

  function openMenu() {
    toggleNavState();
    enableTouchEvents();
    navIndicator.innerHTML = closeIconMarkup;
  }

  function toggleNavState() {
    content.classList.toggle("fade");
    mainNav.classList.toggle("open");
    isOpen = !isOpen;
  }

  function handleNavButtonClick() {
    if (isOpen) {
      closeMenu();
    } else {
      openMenu();
    }
  }

  navButton.addEventListener("click", handleNavButtonClick);
});
