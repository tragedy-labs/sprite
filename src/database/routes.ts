export enum Routes {
  QUERY = 'query',
  COMMAND = 'command',
  BEGIN = 'begin',
  COMMIT = 'commit',
  ROLLBACK = 'rollback',
  EXISTS = 'exists'
}
