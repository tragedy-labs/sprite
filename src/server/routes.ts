export enum Routes {
  READY = 'ready',
  COMMAND = 'server',
  DATABASES = 'databases',
  EXISTS = 'exists'
}
