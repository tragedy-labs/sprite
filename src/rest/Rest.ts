import { SpriteTransaction } from '../transaction/SpriteTransaction.js';
import { HttpClient } from './HttpClient.js';
import { JsonResponse } from './JsonResponse.js';
import { SpriteHeaders } from './SpriteHeaders.js';
import { SpriteBody, SpriteRestBody } from './SpriteBody.js';
import { Routes as DatabaseRoutes } from '../database/routes.js';
import { Routes as ServerRoutes } from '../server/routes.js';
import { Session } from '../session/Session.js';

type ApiRoutes = DatabaseRoutes | ServerRoutes | string;

enum Method {
  POST = 'POST',
  GET = 'GET'
}

/**
 * Static methods for making RESTful API calls.
 */
class Rest {
  /**
   * Make a {@link https://developer.mozilla.org/en-US/docs/Web/HTTP/Methods/GET `GET`} request.
   * @param route The route (key) to target in the request.
   * @param session The `Session` to target in the request.
   * @returns A promise that resolves to a {@link https://developer.mozilla.org/en-US/docs/Web/API/Response `Response`} object.
   */
  public static async get(
    route: ApiRoutes,
    session: Session
  ): Promise<Response> {
    return HttpClient.request(session.endpoints[route], {
      method: Method.GET,
      headers: SpriteHeaders.compose(session),
      keepalive: true
    });
  }
  /**
   * Make a {@link https://developer.mozilla.org/en-US/docs/Web/HTTP/Methods/GET `GET`} request in which a JSON response is expected.
   * @param route The route (key) to target in the request.
   * @param session The `Session` to target in the request.
   * @returns The json response from the ArcadeDB Server.
   */
  public static async getJson<T>(
    route: ApiRoutes,
    session: Session
  ): Promise<T> {
    return JsonResponse.parse<T>(await this.get(route, session));
  }
  /**
   * Make a {@link https://developer.mozilla.org/en-US/docs/Web/HTTP/Methods/POST `POST`} request.
   * @param session The `Session` to target in the request.
   * @param endpoint The relative endpoint of request.
   * @param body The {@link https://developer.mozilla.org/en-US/docs/Web/API/Request/body body} to include in the request.
   * @param headers The {@link https://developer.mozilla.org/en-US/docs/Web/API/Request/headers headers} to include in the request.
   * @returns A promise that resolves to a {@link https://developer.mozilla.org/en-US/docs/Web/API/Response `Response`} object.
   */
  public static async post(
    route: ApiRoutes,
    body: SpriteRestBody,
    session: Session,
    transaction?: SpriteTransaction
  ): Promise<Response> {
    return HttpClient.request(session.endpoints[route], {
      method: Method.POST,
      headers: SpriteHeaders.compose(session, transaction),
      body: body ? SpriteBody.compose(body) : null,
      keepalive: true
    });
  }
  /**
   * Make a {@link https://developer.mozilla.org/en-US/docs/Web/HTTP/Methods/POST `POST`} request in which a JSON response is expected.
   * @param route The route (key) to target in the request.
   * @param body The {@link https://developer.mozilla.org/en-US/docs/Web/API/Request/body body} to include in the request.
   * @param session The `DatabaseSession` to target in the request.
   * @param transaction The `SpriteTransaction` to target in the request.
   * @returns The json response from the ArcadeDB Server.
   */
  public static async postJson<T>(
    route: ApiRoutes,
    body: SpriteRestBody,
    session: Session,
    transaction?: SpriteTransaction
  ): Promise<T> {
    return JsonResponse.parse<T>(
      await this.post(route, body, session, transaction)
    );
  }
}

export { Rest };
