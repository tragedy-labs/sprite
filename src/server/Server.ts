import { ServerSession } from '../session/ServerSession.js';
import { Rest } from '../rest/Rest.js';
import { Routes } from './routes.js';
import { SpriteDatabase } from '../database/SpriteDatabase.js';
import { ArcadeValidation } from '../validation/ArcadeValidation.js';
import {
  type IArcadeCreateUser,
  CLOSE_DATABASE,
  CONNECT_CLUSTER,
  CREATE_DATABASE,
  CREATE_USER,
  DISCONNECT_CLUSTER,
  DROP_DATABASE,
  DROP_USER,
  GET_SERVER_EVENTS,
  OPEN_DATABASE,
  SHUTDOWN
} from './commands/index.js';
import { ArcadeFetchError } from '../errors/ArcadeFetchError.js';

enum CommandResponse {
  OK = 'ok'
}

/**
 * The level of detail that should be returned for a
 * `SpriteServer.getInformation()` request
 * @default 'default'
 */
export type ArcadeServerInformationLevel = 'basic' | 'default' | 'cluster';

/**
 * The metadata returned from the server on every operation.
 */
export type ArcadeServerBasicInformation = {
  /** The user who performed the operation */
  user: string;
  /** The version of ArcadeDB on the server */
  version: string;
  /** The name of the server */
  serverName: string;
};

export type ArcadeServerInfoSetting = {
  key: string;
  value: unknown;
  description: string;
  overridden: boolean;
  default: unknown;
};

export type ArcadeServerInfoMetrics = {
  meters: object;
  events: { errors: number; warnings: number; info: number; hints: number };
  profiler: {
    readCacheUsed: object;
    writeCacheUsed: object;
    cacheMax: object;
    pagesRead: object;
    pagesWritten: object;
    pagesReadSize: object;
    pagesWrittenSize: object;
    pageFlushQueueLength: object;
    asyncQueueLength: object;
    asyncParallelLevel: object;
    pageCacheHits: object;
    pageCacheMiss: object;
    totalOpenFiles: object;
    maxOpenFiles: object;
    walPagesWritten: object;
    walBytesWritten: object;
    walTotalFiles: 8;
    concurrentModificationExceptions: object;
    txCommits: object;
    txRollbacks: object;
    createRecord: object;
    readRecord: object;
    updateRecord: object;
    deleteRecord: object;
    queries: object;
    commands: object;
    scanType: object;
    scanBucket: object;
    iterateType: object;
    iterateBucket: object;
    countType: object;
    countBucket: object;
    evictionRuns: object;
    pagesEvicted: object;
    readCachePages: object;
    writeCachePages: object;
    indexCompactions: object;
    diskFreeSpace: object;
    diskTotalSpace: object;
    diskFreeSpacePerc: object;
    gcTime: object;
    ramHeapUsed: object;
    ramHeapMax: object;
    ramHeapAvailablePerc: object;
    ramOsUsed: object;
    ramOsTotal: object;
    cpuLoad: object;
    jvmSafePointTime: object;
    jvmSafePointCount: object;
    jvmAvgSafePointTime: object;
    totalDatabases: object;
    cpuCores: object;
    configuration: object;
  };
};

export type ArcadeServerDefaultInformation = ArcadeServerBasicInformation & {
  metrics: ArcadeServerInfoMetrics;
  settings: ArcadeServerInfoSetting[];
};

export type ArcadeServerClusterInformation = object;

export type ArcadeServerInformation<T extends ArcadeServerInformationLevel> = {
  basic: ArcadeServerBasicInformation;
  default: ArcadeServerDefaultInformation;
  cluster: ArcadeServerClusterInformation;
}[T];

/**
 * Parameters necessary to create a user in ArcadeDB
 */
export interface ISpriteCreateArcadeUser {
  /** The username of the database user to create */
  username: string;
  /** The password of the database user to create */
  password: string;
  /**
   * An object containing databases to add the user to, and the
   * permisisions to grant them \
   * (i.e. `{ myDatabase: "ADMIN" }`)
   */
  databases: Record<string, string>;
}

/** Describes an event on the ArcadeDB server */
export type SpriteArcadeServerEvent = {
  time: string;
  type: string;
  message: string;
};

/** The ArcadeDB Server Events Log */
export type SpriteArcadeServerEvents = {
  /** The list of server events */
  events: SpriteArcadeServerEvent[];
  /** Event log files */
  files: string[];
};

/**
 * Static methods to interact with an ArcadeDB server.
 */
class Server {
  /**
   * Returns a `boolean` value indicating if the ArcadeDB server is ready.\
   * Useful for remote monitoring of server readiness.
   * @param session The session to use to check the server status.
   * @returns `true` if the server is ready, otherwise `false`.
   * @throws `Error` if the server status could not be checked.
   */
  public static serverReady = async (
    session: ServerSession
  ): Promise<boolean> => {
    try {
      const response = await Rest.get(Routes.READY, session);
      return response.status === 204;
    } catch (error) {
      throw new Error('Unable to check the server status.', { cause: error });
    }
  };
  /**
   * Close a database on the ArcadeDB server.
   * @param session The session to use to close the database.
   * @param databaseName The name of the database to close.
   * @returns The response from the server.
   * @throws `Error` if the database could not be closed.
   */
  public static closeDatabase = async (
    session: ServerSession,
    databaseName: string
  ): Promise<boolean> => {
    try {
      return await this._booleanCommand(session, CLOSE_DATABASE(databaseName));
    } catch (error) {
      throw new Error(`Unabled to close database "${databaseName}`, {
        cause: error
      });
    }
  };
  /**
   * Open a database on the ArcadeDB server
   * @param session The session to use to open the database.
   * @param databaseName The name of the database to open.
   * @returns `true` if the database was opened.
   * @throws `Error` if the database could not be opened.
   */
  public static openDatabase = async (
    session: ServerSession,
    databaseName: string
  ): Promise<boolean> => {
    try {
      return await this._booleanCommand(session, OPEN_DATABASE(databaseName));
    } catch (error) {
      throw new Error(`Unabled to open database "${databaseName}`, {
        cause: error
      });
    }
  };
  /**
   * Returns an `SpriteDatabase` instance for the supplied `databaseName`,
   * using the authorization details of the `Session` instance.
   * @param session The session to use to create the database client.
   * @param databaseName The name of the database to create a client for.
   * @returns An instance of `SpriteDatabase`.
   * @throws `Error` if the database could not be created.
   */
  public static database = (
    session: ServerSession,
    databaseName: string
  ): SpriteDatabase => new SpriteDatabase({ session: session, databaseName });
  /**
   * Sends a command to the ArcadeDB server and returns the response.
   * @param session The session to use to send the command.
   * @param command The command to send to the server.
   * @returns The response from the server.
   * @throws `Error` if the command could not be executed.
   */
  public static command = async <T>(
    session: ServerSession,
    command: string
  ): Promise<T> => {
    try {
      return await Rest.postJson(
        Routes.COMMAND,
        JSON.stringify({
          command
        }),
        session
      );
    } catch (error) {
      throw new Error(
        `There was an error when attemping to execute a command on the ArcadeDB Server. See aggregate error for details.`,
        { cause: error }
      );
    }
  };
  /**
   * Internal method for sending commands to the server in which a JSON response
   * containing an `ok` value in the `result` property is expected. `ok` is then
   * returned as a simple boolean (`true`) value
   * @param session The session to use to send the command.
   * @param command The [command](https://docs.arcadedb.com/#HTTP-ServerCommand) to send to the server, such as `CREATE DATABASE`.
   * @returns `true` if the command was successful.
   * @throws `Error` if the command could not be executed.
   */
  private static _booleanCommand = async (
    session: ServerSession,
    command: string
  ): Promise<boolean> => {
    const response = await this.command(session, command);
    if (response === CommandResponse.OK) {
      return true;
    } else {
      return false;
    }
  };
  /**
   * Connects this server to a cluster with `address`.
   * @param session The session to use to connect to the cluster.
   * @param address The address of the cluster to connect (i.e. 192.168.0.1)
   * @returns The response from the server.
   * @throws `Error` if the cluster could not be connected.
   */
  public static connectCluster = async (
    session: ServerSession,
    address: string
  ): Promise<boolean> => {
    try {
      return await this._booleanCommand(session, CONNECT_CLUSTER(address));
    } catch (error) {
      throw new Error(
        `There was an error attempting to connect cluster at: ${address}`,
        { cause: error }
      );
    }
  };
  /**
   * Create a database
   * @param session The session to use to create the database.
   * @param databaseName The name of the database to create.
   * @returns An instance of `SpriteDatabase`, targeting the created database.
   * @throws `Error` if the database could not be created.
   */
  public static createDatabase = async (
    session: ServerSession,
    databaseName: string
  ) => {
    try {
      ArcadeValidation.databaseName(databaseName);
      const created = await this._booleanCommand(
        session,
        CREATE_DATABASE(databaseName)
      );
      if (created) {
        return this.database(session, databaseName);
      } else {
        throw new Error(
          `Received an unexpected response from the server when attempting to create database "${databaseName}"`
        );
      }
    } catch (error) {
      throw new Error(`Failed to create database "${databaseName}".`, {
        cause: error
      });
    }
  };
  /**
   * Create a user. `username`, `password`, and access controls to multiple databases
   * can be established using the `databases` property of the input parameters.
   * @param session The session to use to create the user.
   * @param username The `username` of the user to create.
   * @param password The `password` of the user to create.
   * @param databases An object of databases to add the user to, and their permissions (groups they belong to).
   * @returns `true` if the user was created successfully.
   * @throws `Error` if the user could not be created.
   */
  public static createUser = async (
    session: ServerSession,
    params: ISpriteCreateArcadeUser
  ): Promise<boolean> => {
    try {
      if (
        !Object.prototype.hasOwnProperty.call(params, 'username') ||
        !Object.prototype.hasOwnProperty.call(params, 'password') ||
        !Object.prototype.hasOwnProperty.call(params, 'databases')
      ) {
        throw new TypeError(
          `The object supplied as an argument must contain 'username', 'password', and 'databases' properties. Received: ${JSON.stringify(
            params
          )}`
        );
      }

      // TODO:
      // ArcadeDB by default requires 4 character minimum length for passwords.
      // The problem is they return a 403 if it's under 4 characters. That could
      // be confusing because it isn't really explained and could be confused with a
      // 403 for invalid credentials from user error establishing authorization
      // through the client. Further question: the manual indicates that it's a 8
      // character minimum, but in practice it's 4 (for non-root users)
      if (!(params.password.length > 3)) {
        throw new TypeError(
          `The password must be at least 4 characters in length, received: ${params.password}, which is ${params.password.length} characters long.`
        );
      }

      // ArcadeDB uses a 'name' property for the user account.
      // This is fine if all you are collecting is a 'username',
      // but in many use cases a user's 'name' and their 'username'
      // will be different fields, and to avoid complication for
      // end users, we are collecting 'username' and manually
      // changing it to 'name' as arcadedb expects.
      const expectedParameters: IArcadeCreateUser = {
        name: params.username,
        password: params.password,
        databases: params.databases
      };

      return await this._booleanCommand(
        session,
        CREATE_USER(expectedParameters)
      );
    } catch (error) {
      const databaseListString = Object.keys(params.databases).join(', ');
      throw new Error(
        `Could not create user ${params.username}. In database(s): ${databaseListString}`,
        { cause: error }
      );
    }
  };
  /**
   * Disconnects the server from the cluster.
   * @param session The session to use to disconnect from the cluster.
   * @returns The response from the server.
   * @throws `Error` if the cluster could not be disconnected.
   */
  public static disconnectCluster = async (
    session: ServerSession
  ): Promise<boolean> => {
    try {
      return await this._booleanCommand(session, DISCONNECT_CLUSTER);
    } catch (error) {
      throw new Error(
        'There was an error when attempting to disconnect from the cluster.',
        { cause: error }
      );
    }
  };
  /**
   * Drop a database
   * @param session The session to use to drop the database.
   * @param databaseName The name of the database to drop.
   * @returns `true` if successfully dropped.
   * @throws `Error` if the database could not be dropped.
   */
  public static dropDatabase = async (
    session: ServerSession,
    databaseName: string
  ): Promise<boolean> => {
    try {
      return await this._booleanCommand(session, DROP_DATABASE(databaseName));
    } catch (error) {
      throw new Error(`Failed to drop database.`, {
        cause: error
      });
    }
  };
  /**
   * Drop a user from the ArcadeDB server.
   * @param session The session to use to drop the user.
   * @param username The `username` of the user to drop from the ArcadeDB server.
   * @returns `true` if the user was successfully dropped.
   * @throws `Error` if the user could not be dropped.
   */
  public static dropUser = async (
    session: ServerSession,
    username: string
  ): Promise<boolean> => {
    try {
      return await this._booleanCommand(session, DROP_USER(username));
    } catch (error) {
      throw new Error(`Could not drop user ${username}.`, { cause: error });
    }
  };
  /**
   * Retrieves a list of server events, optionally a filename of the form
   * `server-event-log-yyyymmdd-HHMMSS.INDEX.jsonl` (where INDEX is a integer, i.e. 0)
   * can be given to retrieve older event logs.
   * @param session The session to use to retrieve the server events.
   * @returns An object containing he server events from the server, and filenames of the associated logs.
   * @throws `Error` if there was a problem fetching the event logs.
   */
  public static getEvents = async (
    session: ServerSession
  ): Promise<SpriteArcadeServerEvents> => {
    try {
      return await this.command<SpriteArcadeServerEvents>(
        session,
        GET_SERVER_EVENTS
      );
    } catch (error) {
      throw new Error(
        `There was an error when attempting to retrieve ArcadeDB server event logs.`,
        { cause: error }
      );
    }
  };
  /**
   * Returns the current configuration.
   * @param session The session to use to retrieve the configuration.
   * @param mode The level of informatio detail to return.
   * * `basic` returns minimal server information
   * * `default` returns full server configuration (default value when no parameter is given)
   * * `cluster` returns the cluster layout
   * @returns The server information.
   * @throws `Error` if the server information could not be retrieved.
   */
  public static getInformation = async <
    M extends ArcadeServerInformationLevel = 'default'
  >(
    session: ServerSession,
    mode?: M
  ): Promise<ArcadeServerInformation<M>> => {
    try {
      // TODO: There should be a method to modify the URL
      // in the Rest class with parameters like this, for now this
      // works
      // TODO: The repetition of the RequestInit is less
      // than ideal
      const response = await fetch(
        `${session.address}/api/v1/server?mode=${mode || 'default'}`,
        {
          method: 'GET',
          headers: session.headers,
          keepalive: true
        }
      );

      switch (response.status) {
        case 200: {
          const jsonResponse = await response.json();
          return jsonResponse.result;
        }
        case 403:
        case 404:
        case 500:
        default:
          throw new ArcadeFetchError(response);
      }
    } catch (error) {
      throw new Error(`Could not get ArcadeDB server information.`, {
        cause: error
      });
    }
  };
  /**
   * Returns a list of database names that are present on the server.
   * @param session The session to use to retrieve the database list.
   * @returns {Promise<Array<string>>} A list (array) of database names present on the server.
   * @throws `Error` if the database list could not be retrieved.
   */
  public static listDatabases = async (
    session: ServerSession
  ): Promise<Array<string>> => {
    try {
      return await Rest.getJson(Routes.DATABASES, session);
    } catch (error) {
      throw new Error(
        'Encountered an error when attemping to fetch list of databases from the server.',
        { cause: error }
      );
    }
  };
  /**
   * Gracefully shutdown the server.\
   * `TODO:` This works, in that it does shutdown the server, but the fetch throws
   * before it resolves, guessing because the server is shutting down. A CURL, however,
   * returns an empty `204` response as the documentation indicates.
   * @param session The session to use to shutdown the server.
   * @returns `true` if the server is successfully shutdown.
   * @throws `Error` if there is a problem attempting the shutdown.
   */
  public static shutdown = async (session: ServerSession): Promise<boolean> => {
    try {
      const response = await Rest.post(
        Routes.COMMAND,
        { command: SHUTDOWN },
        session
      );
      return response.status === 204;
    } catch (error) {
      throw new Error(
        `There was an error when attempting to shutdown the ArcadeDB server at.`,
        { cause: error }
      );
    }
  };
}

export { Server };
