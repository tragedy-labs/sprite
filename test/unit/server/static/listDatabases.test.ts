import { Rest } from '@/rest/Rest.js';
import { Routes } from '@/server/routes.js';
import { Server } from '@/server/Server.js';

import { TestServerSession as SESSION } from '@test/variables.js';

describe('Server.listDatabases()', () => {
  beforeEach(() => {
    jest.clearAllMocks(); // Clear mocks before each test to avoid interference
  });
  it('should make a properly formatted fetch request with supplied options', async () => {
    const expectedResponse = ['db1', 'db2'];

    // Mocking Rest.getJson
    jest.spyOn(Rest, 'getJson').mockResolvedValue(expectedResponse);

    const result = await Server.listDatabases(SESSION);

    // Assert
    expect(Rest.getJson).toHaveBeenCalledWith(Routes.DATABASES, SESSION);
    expect(result).toEqual(expectedResponse);
  });

  it('should handle errors thrown by Rest.getJson and rethrow them', async () => {
    const error = new Error('Network error');

    // Mocking Rest.getJson to throw an error
    jest.spyOn(Rest, 'getJson').mockRejectedValue(error);

    await expect(Server.listDatabases(SESSION)).rejects.toThrow(
      'Encountered an error when attemping to fetch list of databases from the server.'
    );
  });
});
