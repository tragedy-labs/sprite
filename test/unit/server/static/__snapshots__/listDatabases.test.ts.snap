// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SpriteServer.listDatabases() should return an error for a 403 status 1`] = `[Error: Encountered an error when attemping to fetch list of databases from the server.]`;
